# 2-D Convolution

# Question 1
### Build your own “2-D Convolution” function file to perform the convolution operation, given the input matrix of size m x m and one filter of size f x f, such that the ouput convolution length is given by m-f+1.

<img src = "2dconv.png">

#### In the above example (a), the size of the input is 6X6 and size of filter is 3X3. So the output size is 6-3+1 => 4X4

In [129]:
def conv2d(inp, filt, b):
    t = 0
    temp = 0
    while(t < len(inp)):
        temp = temp + np.dot(inp[t].T, filt[t])
        t = t + 1
    
    return temp+len(filt)*len(filt)*b

In [130]:
import numpy as np
np.random.seed(52)

## mXm - size of input
## fXf - size of filter

In [131]:
m = np.random.randint(5,15) # Input size
f = (np.random.randint(1,2)*2+1) # Filter size

In [132]:
m,f

(10, 3)

#### Since we got m = 10 and f = 3, the output size should be 8X8

In [133]:
inp1 = np.random.randint(-5,5, size = (m,m))
filt1 = np.random.randint(-5,5, size = (f,f))
bias = np.random.randint(-5,5)

In [134]:
inp1,len(inp1),filt1,bias

(array([[ 2,  1,  2, -5,  0, -2, -2, -4, -2, -4],
        [ 0, -4, -5, -5,  1,  1,  2,  4, -1,  1],
        [ 2,  2,  2, -1, -3,  0, -1, -5, -4, -3],
        [ 0, -3,  1,  4,  1,  0, -5,  1,  3, -4],
        [ 4,  2, -4,  4,  1,  4,  4, -4,  0,  3],
        [-1,  4, -5,  1,  2, -2,  2, -5, -5,  4],
        [ 0, -4,  4,  2,  1,  4,  2, -1,  2, -5],
        [-4,  0, -4,  4, -1, -5,  4,  1, -2, -4],
        [ 3,  2,  1, -1,  1,  0,  3, -4, -5, -5],
        [ 2, -3, -5, -3, -3, -2, -1,  2, -5, -4]]),
 10,
 array([[-5, -5, -2],
        [-2, -1,  0],
        [-4, -1, -4]]),
 4)

#### Now, I'll do 2-D convolution and the function is defined above and I use same bias value for all the problems

In [135]:
conv1 = np.zeros([m-f+1,m-f+1])
for i in range(0,m-f+1):
    for j in range(0,m-f+1):
        #print(inp1[i:i+3,j:j+3])
        conv1[i,j] = conv2d(inp1[i:i+3,j:j+3], filt1, bias)
conv1,conv1.shape

(array([[  3.,  38.,  71.,  81.,  63.,  81.,  87., 103.],
        [ 59.,  80.,  69.,  42.,  44.,  15.,  22.,  42.],
        [ 13.,   3.,  39.,  14.,  27.,  52.,  71.,  86.],
        [ 59.,  23.,  24.,   4.,  21.,  73.,  63.,  41.],
        [  0.,  39.,  21., -26., -15.,  -8.,  22.,  87.],
        [ 67.,  31.,  58.,  25.,  19.,  48.,  49.,  92.],
        [ 38.,  31.,   1.,   9.,  -2.,  27.,  30.,  82.],
        [ 71.,  72.,  72.,  55.,  74.,  37.,  35.,  75.]]),
 (8, 8))

#### So we can see the output of the 2-D convolution operation and we got the size 8X8

# Question 2
### Build your own “2-D Convolution” function file to perform the convolution operation given the input matrix of size m x m and one filter of size f x f,, such that the ouput convolution length is same as that of the length of the input sequence.

#### Since the output should be the same size as that of input, we have to padd 0's to the input and perform the convolution
#### And here I'm taking the same input as that of the 1st question and padd the 0's and same filter

    If mXm (10X10) is input size, and filter size is fXf (3X3) and ouput size should be nXn (m=n)
    Then,
           m-f+1+2x = n
           2x = (n-m)+f-1
           10-3+1+2x = 10
        => x = 1
    That means, I've to padd 0's as 1-layer around the input
    So the resultant input size is,   m+2x  ,     where 2x = (n-m)+f-1
                                    = 12X12

In [136]:
n=m
npad = int((m-n+f-1))
#inp2 = np.random.randint(-5, 5 , size = ([14,14]))
#m = len(inp2)
inp2 = np.zeros([m+npad,m+npad])
filt2 = filt1
inp2[int(npad/2):m+int(npad/2),int(npad/2):m+int(npad/2)] = inp1

In [137]:
inp2,filt2

(array([[ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
        [ 0.,  2.,  1.,  2., -5.,  0., -2., -2., -4., -2., -4.,  0.],
        [ 0.,  0., -4., -5., -5.,  1.,  1.,  2.,  4., -1.,  1.,  0.],
        [ 0.,  2.,  2.,  2., -1., -3.,  0., -1., -5., -4., -3.,  0.],
        [ 0.,  0., -3.,  1.,  4.,  1.,  0., -5.,  1.,  3., -4.,  0.],
        [ 0.,  4.,  2., -4.,  4.,  1.,  4.,  4., -4.,  0.,  3.,  0.],
        [ 0., -1.,  4., -5.,  1.,  2., -2.,  2., -5., -5.,  4.,  0.],
        [ 0.,  0., -4.,  4.,  2.,  1.,  4.,  2., -1.,  2., -5.,  0.],
        [ 0., -4.,  0., -4.,  4., -1., -5.,  4.,  1., -2., -4.,  0.],
        [ 0.,  3.,  2.,  1., -1.,  1.,  0.,  3., -4., -5., -5.,  0.],
        [ 0.,  2., -3., -5., -3., -3., -2., -1.,  2., -5., -4.,  0.],
        [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.]]),
 array([[-5, -5, -2],
        [-2, -1,  0],
        [-4, -1, -4]]))

#### Now the input size is 12X12 , filter size os 3X3. So the output size should be 10X10 (Actual input size)

In [138]:
conv2 = np.zeros([m+npad-f+1,m+npad-f+1])
for i in range(0,len(inp2)-f+1):
    for j in range(0,len(inp2[0])-f+1):
        #print(inp1[i:i+3,j:j+3])
        conv2[i,j] = conv2d(inp2[i:i+f,j:j+f], filt2, bias)
conv2,conv2.shape

(array([[ 50.,  55.,  73.,  58.,  61.,  25.,  20.,  36.,  27.,  47.],
        [ 14.,   3.,  38.,  71.,  81.,  63.,  81.,  87., 103.,  86.],
        [ 54.,  59.,  80.,  69.,  42.,  44.,  15.,  22.,  42.,  39.],
        [ 10.,  13.,   3.,  39.,  14.,  27.,  52.,  71.,  86.,  66.],
        [ 23.,  59.,  23.,  24.,   4.,  21.,  73.,  63.,  41.,  54.],
        [ 29.,   0.,  39.,  21., -26., -15.,  -8.,  22.,  87.,  24.],
        [ 37.,  67.,  31.,  58.,  25.,  19.,  48.,  49.,  92.,  54.],
        [ 37.,  38.,  31.,   1.,   9.,  -2.,  27.,  30.,  82.,  84.],
        [ 63.,  71.,  72.,  72.,  55.,  74.,  37.,  35.,  75., 105.],
        [ 15.,   8.,  34.,  47.,  45.,  33.,  34.,  51.,  92., 100.]]),
 (10, 10))

#### Here we can see that the output size is 10X10. We can given any size input and get thge correspoinf output

# Question 3
### Use any one of your built “Conv2” function file to perform the convolution operation given the input matrix and ‘N1’ number of filters.

#### Here, I can take an input matrix with size mXm and no. of filters as nf with each filer size as fXf
#### So the total output size should be nf X (m-f+1) X (m-f+1)
#### So for this question, I'm taking another random input

In [139]:
np.random.seed(51)
m = np.random.randint(5,15) # Size of input
nf = np.random.randint(2,5) # No. of filters
f = np.random.randint(2,4)*2+1 # Filter size
filt3 = np.random.randint(-5,5, size = (nf,f,f))
inp3 = np.random.randint(-5,5, size = (m,m))

In [140]:
inp3,inp3.shape,filt3,filt3.shape,nf,f

(array([[-5, -1,  0, -1,  4, -3,  3,  0,  1, -3,  2,  2, -1,  0],
        [-3,  4, -4, -1, -2,  1,  4, -3,  1, -5,  3, -5, -5,  1],
        [-4,  0, -4, -4,  4,  1, -1,  0, -3,  1,  0, -5, -2,  0],
        [-3,  3,  1, -3, -2, -3, -3,  3, -1, -2, -5, -2,  4, -2],
        [ 1,  1,  3,  1, -5, -4, -3, -5,  3,  1,  3,  1,  3, -3],
        [-5,  0,  0,  3, -2, -5,  1, -1, -2,  0,  3,  3,  0,  3],
        [-3,  1, -4, -5, -4, -5, -1,  3, -2,  1, -4,  0,  0, -2],
        [-4, -5,  0, -1, -5, -1,  2,  3, -2, -3,  2,  4,  0, -2],
        [ 3,  4,  1, -1,  0,  0,  3,  4, -4, -4,  3, -2, -5, -2],
        [-3,  2, -5, -4,  3, -5, -4, -4,  1, -3,  2,  3, -1,  2],
        [ 0,  2,  1, -5, -1,  0, -5,  0,  4,  2, -3,  0,  0, -3],
        [-2, -1,  2,  3,  4, -1, -5, -3, -3, -4, -2, -2, -5,  1],
        [-4, -5, -4,  3,  4,  0,  3,  4, -5, -3, -4,  3,  3, -4],
        [ 3, -5,  4, -3,  0,  0, -1,  3,  4, -2,  3,  1, -5,  3]]),
 (14, 14),
 array([[[ 4,  0, -5,  0, -1],
         [ 4,  0,  0, -1, -1],
 

##### Here, we can see that the input size is 14X14 (mXm) and filter size 5X5 with 3 filters
#### So
    We shout get the output as nf X (m-f+1) X (m-f+1)
                                i.e 3 X 10 X 10

In [141]:
conv3 = np.zeros([nf,m-f+1,m-f+1])
for i in range(0,nf):
    for j in range(0,m-f+1):
        for k in range(0,m-f+1):
            conv3[i,j,k] = conv2d(inp3[j:j+f,k:k+f], filt3[i], bias)

In [142]:
conv3,conv3.shape

(array([[[ 92., 100.,  57., 130.,  59., 111., 126.,  74., 140.,  39.],
         [ 49.,  84.,  53.,  96., 125., 105., 134., 119.,  96.,  90.],
         [ 49., 100.,  63.,  11., 120., 123., 105., 144.,  87., 113.],
         [ 77.,  95., 187., 139.,  51.,  89.,  92., 153., 117.,  95.],
         [ 86., 139., 158., 182.,  71.,  90.,  98.,  41.,  48.,  82.],
         [ 62.,  71.,  78., 109.,  60.,  39., 112.,  85.,  87., 110.],
         [124.,  85., 135., 109.,  81.,  49., 141., 110.,  81., 114.],
         [152.,  58.,  75., 131.,  75.,  61., 123., 163.,  10.,  51.],
         [101.,  95.,  50., 115.,  97.,  66., 143., 149.,  78.,  74.],
         [146., 145.,  87.,  64.,  76.,  87.,  41., 127.,  89.,  92.]],
 
        [[ 73., 118.,  91., 141.,  91., 154., 131., 135., 110.,  -1.],
         [ 76.,  90.,  97.,  97., 161.,  97.,  62., 122.,  82.,  83.],
         [ 71., 118., 172.,  96., 120., 131., 103., 125.,  94., 129.],
         [196., 109., 112., 136.,  89., 108.,  89., 102., 128., 134.],
   

# Question 4
### Use any one of your built “Conv2” function file to perform the convolution operation given the input matrix and ‘N1’ number of filters, followed by the convolution operation with ‘N2’ number of filters, where N2>N1.

#### For this one, I'll take the output of the previous question and give it as the input to the next layers eith filters N2 where N2 > 3(nf)
#### And let's the size of each filter is less than f(5)

In [150]:
np.random.seed(51)
f1 = np.random.randint(3,5) # Filter size
nf1 = np.random.randint(nf+1, nf+5) # No. of filters
inp4 = conv3 # Input to the next level
filt4 = np.random.randint(-5,5, size = (nf1, nf, f1, f1)) # Filter

In [151]:
[no,m,n]=conv3.shape
inp4,inp4.shape,filt4,filt4.shape,nf1,f1

(array([[[ 92., 100.,  57., 130.,  59., 111., 126.,  74., 140.,  39.],
         [ 49.,  84.,  53.,  96., 125., 105., 134., 119.,  96.,  90.],
         [ 49., 100.,  63.,  11., 120., 123., 105., 144.,  87., 113.],
         [ 77.,  95., 187., 139.,  51.,  89.,  92., 153., 117.,  95.],
         [ 86., 139., 158., 182.,  71.,  90.,  98.,  41.,  48.,  82.],
         [ 62.,  71.,  78., 109.,  60.,  39., 112.,  85.,  87., 110.],
         [124.,  85., 135., 109.,  81.,  49., 141., 110.,  81., 114.],
         [152.,  58.,  75., 131.,  75.,  61., 123., 163.,  10.,  51.],
         [101.,  95.,  50., 115.,  97.,  66., 143., 149.,  78.,  74.],
         [146., 145.,  87.,  64.,  76.,  87.,  41., 127.,  89.,  92.]],
 
        [[ 73., 118.,  91., 141.,  91., 154., 131., 135., 110.,  -1.],
         [ 76.,  90.,  97.,  97., 161.,  97.,  62., 122.,  82.,  83.],
         [ 71., 118., 172.,  96., 120., 131., 103., 125.,  94., 129.],
         [196., 109., 112., 136.,  89., 108.,  89., 102., 128., 134.],
   

In [173]:
conv4 = np.zeros([nf1,m-f1+1,n-f1+1])
conv4.shape

(5, 8, 8)

In [174]:
temp = np.zeros([m-f1+1,n-f1+1])
summ = np.zeros([m-f1+1,n-f1+1])

t = 0
for i in filt4:
    for j in range(0,len(inp4)):
        for k in range(0,m-f1+1):
            for l in range(0,n-f1+1):
                temp[k,l] = conv2d(inp4[j,k:k+f1,l:l+f1], i[j], bias)
        summ = summ + temp
    conv4[t] = summ
    summ = np.zeros([m-f1+1,n-f1+1])
    t = t + 1

In [176]:
conv4,conv4.shape

(array([[[  627.,   795.,  -504.,    21.,     0.,  -393.,   608.,   198.],
         [  179.,  1186.,  1183.,  -465.,   238., -1245.,  -140.,   -15.],
         [ -142.,  -611.,   799.,   427.,   267.,   152.,  -793.,  -224.],
         [  674.,  -589.,  -963.,   778.,   151.,   918.,   -61.,    67.],
         [  162.,  -346., -1013.,   683.,   737.,  -149.,   215.,   256.],
         [  359.,    75.,  -649.,  -429.,   845.,  -516.,   237.,    18.],
         [ 1342.,   365.,  -185.,  -404.,   417.,   -48.,  -712.,   604.],
         [  813.,  1570.,   225.,    88.,   562.,  -477., -1007.,   602.]],
 
        [[-1526.,  -986.,  -619.,  -793.,  -789., -1968., -2104., -2268.],
         [-1128., -1091., -1136., -1873., -1111., -1194., -1206., -2014.],
         [ -626., -1892., -1715., -1307., -1767., -1357., -1191.,  -751.],
         [-1718., -1414., -2464., -1769., -1401.,  -865., -1338., -1371.],
         [-1497., -1443., -1582., -1592.,  -985., -1202.,  -872.,  -919.],
         [-1501., -101

# Question 5
### Compute the number of learnable parameters for Q.No.4. Validate your computation with the number of learnable parameters given by Keras-Tensorflow framework.

    No. of Filters and size of each filter the 1st layer = 3*5*5 = 75
    And no. of bias values to be added is the no. of filters = 3
    Total no. of parameters w.r.t 1st layer = 78
    
    No. of layers of filters in 2nd layer = 3 and No. of filter = 5
        Total = 15
    Each filter size = 3*3 = 9
    Parametrs = 15*9 = 135
    And no. of bias values to be added is the no. of filters = 5
    Total no. of learnable parameters in the 2nd layer = 140
    
    So, Total no. of learnable parameters with respect to this problem = 140 + 78
                                                                       = 218

In [180]:
print("Total learnable parameters = ", (nf*f*f + nf) + nf1*nf*f1*f1 + nf1)

Total learnable parameters =  218


In [261]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, Flatten, MaxPool1D,Dropout

In [262]:
model = Sequential()
model.add(Conv2D(nf, (f,f), input_shape = (inp3.shape[0],inp3.shape[1],1)))

In [263]:
model.add(Conv2D(nf1, (f1,f1)))

In [264]:
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits = True)

In [265]:
model.compile(optimizer = 'adam', loss = loss_fn, metrics = ['accuracy'])

In [266]:
model.summary()

Model: "sequential_16"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_28 (Conv2D)           (None, 10, 10, 3)         78        
_________________________________________________________________
conv2d_29 (Conv2D)           (None, 8, 8, 5)           140       
Total params: 218
Trainable params: 218
Non-trainable params: 0
_________________________________________________________________
